ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/JmsXAResourceRecoveryHelper.java

```diff
@@ -177,4 +268,11 @@ private void disconnect() throws XAException {
         }
     }
 
+    private void assertIfRecoveryScanWasStarted() throws XAException {
+        if (delegate == null) {
+            LOGGER.warn("XAResource.recover has to be called first with XAResource.TMSTARTRSCAN argument");
+            throw new XAException(XAException.XAER_PROTO);
```

Comment:
Actually I just looked at what we do for XAER_PROTO and XAER_INVAL during recovery and we return HEURISTIC_HAZARD in both cases. I think it should be safe to assume that we have coded up the recovery module to correctly handle the recover calls so I do think we should be ok.